### ISSUE TITLE
Call to function "coalesce" failed: all arguments must have the same type.

### ISSUE THREAD

**Issue by TheodorWu** [Azure/terraform-azurerm-caf-enterprise-scale#340]
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Versions

**terraform**: v1.1.8

**azure provider**: v2.99.0

**module**: v1.1.4 (based on main branch of repository, latest commit: 0b343c6ba9ce4595a0e494f97d2a439845a1cb2a)

### Description

When specifying the advanced block of the `configure_connectivity_resources`  input as follows:
```
advanced = {
      resource_prefix                           = "pre"
      resource_suffix                           = "suf"
      existing_ddos_protection_plan_resource_id = ""
      existing_virtual_wan_resource_id          = ""
      resource_group_per_virtual_hub_location   = false
      custom_settings_by_resource_type = {
        azurerm_resource_group = {
          virtual_wan = {
            (local.default_location) = {
              name = local.connectivity_resource_group_name
            }
          }
      }
    }
  }
```
I encounter an error as shown here. I suspect at some point an unintended type conversion was performed by terraform. 

```
│   Error: Error in function call
│
│   on .terraform\modules\enterprise_scale\locals.tf line 53, in locals:
│   53:   connectivity_resources_advanced = coalesce(local.configure_connectivity_resources.advanced, local.empty_map)
│     ├────────────────
│     │ local.configure_connectivity_resources.advanced is object with 6 attributes
│     │ local.empty_map is object with no attributes
│
│ Call to function "coalesce" failed: all arguments must have the same type.
```

**Steps to Reproduce**

1. Specify the advanced block as shown previously
2. `terraform plan`

**Additional Context**

I managed to solve this issue locally when explicitly casting the `local.empty_map` as type `map `using `tomap()` (`locals.tf`, line 5).

```
# The following block of locals are used to avoid using
# empty object types in the code
locals {
  empty_list   = []
  empty_map    = tomap({})
  empty_string = ""
}
```

**Comment by krowlandson** [comment]
Great catch, and thank you for raising awareness of this issue @TheodorWu 

I have fixed another issue in the latest update (#331) which we identified following our first pass at moving from `try()` to `catch()`, but missed this one. Our fix should prevent the need to declare all of the "top-level" advanced attributes if not in use.

I will run some more tests to check whether explicitly casting the `local.empty_map` as type `map` using `tomap()` (`locals.tf`, line 5) doesn't change any other behaviours but this looks like a good fix which we can quickly roll-out to `main`. Thank you!

![image](https://user-images.githubusercontent.com/12268562/163455618-4b5db6ee-9c0b-48e6-a9c0-d5eeb7f3aec7.png)

![image](https://user-images.githubusercontent.com/12268562/163455649-cefda031-b755-489a-99ee-77469a40d643.png)

![image](https://user-images.githubusercontent.com/12268562/163455852-69633237-6958-4ec2-9290-4bb48272a548.png)
